agent/client/workflow_listen_for_trigger_requests.go

```diff
@@ -22,12 +21,26 @@ func (c *Client) startTriggerListener(ctx context.Context) error {
 		for {
 			resp := proto.TriggerRequest{}
 			err := stream.RecvMsg(&resp)
-			if errors.Is(err, io.EOF) || isCancelledError(err) {
+			if isEndOfFileError(err) || isCancelledError(err) {
 				return
 			}
 
+			if err != nil && isConnectionError(err) {
+				err = retry.Do(func() error {
+					return c.reconnect()
+				})
+				if err == nil {
+					// everything was reconnect, so we can exist this goroutine
+					// as there's another one running in parallel
+					return
+				}
+
+				log.Fatal(err)
+			}
+
```

Comment:
all this functiosn are copy/pasted. can't they be generalizaed? like:

```
func (c Client) handleDisconnectionError(inputErr error) error {
    if inputErr == nil {
        return  nil
    }
	  if isConnectionError(inputErr) {
        return retry.Do(func() error {
					return c.reconnect()
				})
		}
    return inputErr
}
```

and the 
```suggestion
      err = c.handleConnectionError(err)
			if err != nil {
				log.Fatal(err)
			}
```

Reply:
I used the core idea of your suggestion.

I didn't follow it 100% for two reasons:

1. I cannot exit the process if any error happens, only if it's a connection error and we can't connect.
2. I still need to exit the goroutine because when `c.reconnect()` is called, a new goroutine will spawn with a new `stream` object. When reconnecting to a grpc server, we also need to get a new `stream` using the new connection. So I have to have a way of ensuring that the old goroutine will not run again. That's why we have this:

```go
if err == nil {
  return
}
```

It means that the reconnection was sucessfull and now we have a new `grpc.Conn` and new `grpc streams` as well.